lib/src/components/mixins/hitbox.dart

```diff
@@ -1,84 +1,70 @@
+import 'dart:math';
 
-import '../../../components.dart';
-import '../../collision_detection.dart' as collision_detection;
+import '../position_component.dart';
+import '../../../extensions.dart';
+import '../../geometry/shape.dart';
 
 mixin Hitbox on PositionComponent {
-  List<Vector2> _shape;
+  final List<HitboxShape> _shapes = <HitboxShape>[];
 
-  /// The list of vertices used for collision detection and to define whether
-  /// a point is inside of the component or not, so that the tap detection etc
-  /// can be more accurately performed.
-  /// The hitbox is defined from the center of the component and with
-  /// percentages of the size of the component.
-  /// Example: [[1.0, 0.0], [0.0, 1.0], [-1.0, 0.0], [0.0, -1.0]]
-  /// This will form a square with a 45 degree angle (pi/4 rad) within the
-  /// bounding size box.
-  set shape(List<Vector2> vertices) => _shape = vertices;
-  List<Vector2> get shape => _shape ?? [];
+  UnmodifiableListView<HitboxShape> get shapes => UnmodifiableListView(_shapes);
 
-  /// Whether the hitbox shape has defined vertices and is not an empty list
-  bool hasShape() => _shape?.isNotEmpty ?? false;
+  void addShape(HitboxShape shape) {
+    shape.component = this;
+    _shapes.add(shape);
+  }
 
-  Iterable<Vector2> _scaledShape;
-  Vector2 _lastScaledSize;
+  void removeShape(HitboxShape shape) {
+    _shapes.remove(shape);
+  }
 
-  /// Gives back the shape vectors multiplied by the size of the component
-  Iterable<Vector2> get scaledShape {
-    if (_lastScaledSize != size || _scaledShape == null) {
-      _lastScaledSize = size;
-      _scaledShape = _shape?.map(
-        (p) => p.clone()..multiply(size / 2),
-      );
-    }
-    return _scaledShape;
+  /// Checks whether the hitbox represented by the list of [HitboxShape]
+  /// contains the [point].
+  @override
+  bool containsPoint(Vector2 point) {
+    return possiblyContainsPoint(point) &&
+        _shapes.any((shape) => shape.containsPoint(point));
   }
 
-  void renderContour(Canvas canvas) {
-    final hitboxPath = Path()
-      ..addPolygon(
-        scaledShape.map((point) => (point + size / 2).toOffset()).toList(),
-        true,
-      );
-    canvas.drawPath(hitboxPath, debugPaint);
+  void renderShapes(Canvas canvas) {
+    _shapes.forEach((shape) => shape.render(canvas, debugPaint));
   }
 
-  // These variables are used to see whether the bounding vertices cache is
-  // valid or not
-  Vector2 _lastCachePosition;
-  Vector2 _lastCacheSize;
-  double _lastCacheAngle;
-  bool _hadShape = false;
-  List<Vector2> _cachedHitbox;
+  /// Returns the absolute [Rect] that contains all the corners of the rotated
+  /// [toAbsoluteRect] rect.
+  Rect toBoundingRect() {
+    final rotatedPoints = toAbsoluteRect().toVectors()
+      ..forEach((v) => v.rotate(
+            angle,
+            center: absolutePosition,
+          ));
+    final minX =
+        rotatedPoints.map<double>((v) => v.x).fold(double.infinity, min);
```

Comment:
you can use reduce to avoid the double.infinity (here and below)

if the list is empty we wanna throw an error anyway, maybe we can even add that assert to the Polygon constructor

Reply:
Of course, classic map-reduce!

Reply:
fold is a fancier version with a default value for empty lists